### PR TITLE
Python: Use API graphs in PEP249 support

### PR DESCRIPTION
**Pull request by tausbn**
Because the replacement extension point now extends `API::Node`, I
modified the `toString` method of the latter to have an empty body.
The alternative would be to require everyone to provide a `toString`
predicate for their extensions, but seeing as these will usually be
pointing to already existing API graph nodes, this seems silly.

(This may be the reason why the equivalent method in the JS libs has
such an implementation.)

---
This one probably needs a change note, since it deprecates part of the API. I'm not entirely sold on the name `PEP249ModuleApiNode`, so any suggestions for a better name are welcomed.

### CODE REVIEW

File: python/ql/src/semmle/python/frameworks/PEP249.qll [hunk 1]
```diff
@@ -7,20 +7,23 @@ private import python
...
-  t.startInAttr("connect") and
-  result instanceof PEP249Module
-  or
-  exists(DataFlow::TypeTracker t2 | result = connect(t2).track(t2, t))
-}
+/**
+ * An abstract class encompassing API graph nodes that implement PEP 249.
+ * Extend this class for implementations.
+ */
+abstract class PEP249ModuleApiNode extends API::Node { }
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/PEP249.qll:23]: Could we not just add the `toString` method here, and thus not require it in every model?

**Review by yoff — COMMENTED** [review]
I am very happy about how small the models become :-D
I am slightly worried about the potential foot-gun of no toString-values during debugging.

File: python/ql/src/semmle/python/frameworks/PEP249.qll (hunk 1 above)
Comment (tausbn) [reply python/ql/src/semmle/python/frameworks/PEP249.qll:23]: > I am slightly worried about the potential foot-gun of no toString-values during debugging.

Just to be clear, all of the relevant nodes have `toString` representations at the moment. The only time this becomes a potential foot-gun is when we're adding new node types to `API::Impl::TApiNode`. In this case, we might forget to add `toString`, though I think it is unlikely.

> Could we not just add the toString method here, and thus not require it in every model?

We could, but what should be the string returned? All nodes already have appropriate `toString`s... I guess we could prefix them with the fact that they're `PEP249ModuleApiNode`s, but I'm not sure that actually adds value.

Maybe I'm doing something odd in terms of inheritance here? :thinking:

Comment (intrigus-lgtm) [reply python/ql/src/semmle/python/frameworks/PEP249.qll:23]: Would it not be possible to add a test that fails when any API::Impl::TApiNode class does not have a toString method?

Comment (tausbn) [reply python/ql/src/semmle/python/frameworks/PEP249.qll:23]: That's... not a terrible idea. We can't test on `TApiNode` directly, as it's hidden inside a `private` module, but we _could_ do
```ql
select count(API::Node n | not exists(n.toString()))
```
in a test (with expected output `0`) and that should at least warn us if this changes.

File: python/change-notes/2021-04-13-pep249-api-graphs.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+lgtm,codescanning
+* Modelling of libraries supporting PEP249 has been changed to use API graphs. When defining new
```
Comment (intrigus-lgtm) [inline python/change-notes/2021-04-13-pep249-api-graphs.md:2]:
```suggestion
* Modeling of libraries supporting PEP249 has been changed to use API graphs. When defining new
```

Comment (tausbn) [reply python/change-notes/2021-04-13-pep249-api-graphs.md:2]: Curse my British mannerisms! :smile:

**Review by yoff — APPROVED** [review]
LGTM

**Pull request merged**